### PR TITLE
765 location hops fix

### LINKED ISSUE #765: Location management not correctly updating
**What Needs to be Done?**

Messages may be taking too many hops and leading to slow downs.

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #765

### CODE REVIEW

File: src/vt/topos/location/location.impl.h [hunk 1]
```diff
@@ -399,6 +399,53 @@ void EntityLocationCoord<EntityID>::getLocation(
   }
 }
 
+template <typename EntityID>
+void EntityLocationCoord<EntityID>::handleEagerUpdate(
+  EntityID const& id, NodeType home_node, NodeType deliver_node
+) {
+  auto this_node = theContext()->getNode();
+  vtAssert(this_node != deliver_node, "This should have been a forwarding node");
+  vtAssert(home_node != uninitialized_destination, "Home node should be valid");
```
Comment (pnstickne) [inline src/vt/topos/location/location.impl.h:408]: Same effect, plus top-range checking:

     home_node < theContext()->getNumNodes()

File: src/vt/topos/location/location.impl.h [hunk 2]
```diff
@@ -399,6 +399,53 @@ void EntityLocationCoord<EntityID>::getLocation(
...
+  NodeType deliver_node
+) {
+  debug_print(
+    location, node,
+    "sendEagerUpdate: id={}, ask_node={}, home_node={}, deliver_node={}\n",
+    id, ask_node, home_node, deliver_node
+  );
+
+  auto this_node = theContext()->getNode();
+  if (ask_node != this_node) {
```
Comment (pnstickne) [inline src/vt/topos/location/location.impl.h:440]: `!=` or `not_eq`? I'm fine with either, just looking for consistency in new code.

File: tests/unit/location/test_hops.cc [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  }
+
+  void informNeighbor(NeighborMsg* m) {
+    auto proxy = this->getCollectionProxy();
+    auto sidx = m->idx_;
+    proxy(sidx).template send<TestMsg,&TestColl::getBack>(m->do_check_);
+    if (m->do_check_ and m->getHops() > 1) {
+      vt_print(gen, "found long hop message hops={}\n", m->getHops());
+    }
+    vtAssert(
```
Comment (lifflander) [inline tests/unit/location/test_hops.cc:113]: This should be a gtest expect

File: src/vt/topos/location/location.impl.h (hunk 1 above)
Comment (lifflander) [reply src/vt/topos/location/location.impl.h:408]: added

File: src/vt/topos/location/location.impl.h (hunk 2 above)
Comment (lifflander) [reply src/vt/topos/location/location.impl.h:440]: I've been using either style, so I'll keep this as is

File: tests/unit/location/test_hops.cc (hunk 3 above)
Comment (lifflander) [reply tests/unit/location/test_hops.cc]: Fixed

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/topos/location/location.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5357044&bid=17553463)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/topos/location/location.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5357044&bid=17553463)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/topos/location/location.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5357044&bid=17553463)

**Pull request merged**